aopy/analysis/controllers.py

```diff
@@ -0,0 +1,204 @@
+    at each frequency. Assumes output signals are produced by an LTI system.
+    For math details, see Yamagami et al. IEEE Transactions on Cybernetics (2021).
+    
+    Args:
+        input (nt, nch): time-domain input signal
+        output (nt, nch): time-domain output signal
+        samplerate (float): sampling rate of the data
+        exp_freqs ((nfreq,), optional): list or array of frequencies used to generate experimental signals (reference and/or disturbance)
+
+    Returns:
+        tuple: Tuple containing:
+            | **freqs ((nt/2,) or (nfreq,)):** array of frequencies (essentially the x axis of a spectrogram) 
+            | **transfer_func ((nt/2, nch) or (nfreq, nch)):** array of complex numbers corresponding to the transformation at each of the above frequencies
+
+    '''    
+    assert len(input) == len(output), "Mismatched signal lengths"
+
+    # Compute FFT of time-domain signals
+    freqs, freq_input = calc_freq_domain_values(input, samplerate)
+    _, freq_output = calc_freq_domain_values(output, samplerate)
+
+    # Divide freq-domain output by freq-domain input at each frequency to find transformation
+    transfer_func = np.divide(freq_output, freq_input)
+
+    # Find the transformation at specific frequencies of interest
+    if exp_freqs is not None:
+        exp_freq_idx = np.isin(freqs.round(4), exp_freqs)
+        freqs = freqs[exp_freq_idx]
+        transfer_func = transfer_func[exp_freq_idx]
+
+    return freqs.round(4), transfer_func
+
+
+def pair_trials_by_frequency(ref_freqs, dis_freqs, max_trial_distance=3):
+    '''
+    Finds pairs of trials with complementary frequency content (e.g. one trial's task signals consist of a subset
+    of the experimental frequencies, and the other trial's task signals consist of the remaining experimental frequencies). 
+    Assumes that reference and disturbance signals are made up of different frequencies within the same trial.
+
+    Args:
+        ref_freqs (ntrial,): list or array of frequencies used to generate the reference signal for each trial
+        dis_freqs (ntrial,): list or array of frequencies used to generate the disturbance signal for each trial
+        max_trial_distance (int): maximum number of trials allowed between two trials identified as a pair
+
+    Returns:
+        trial_pairs (npair, 2): array of trial indices corresponding to pairs of trials with complementary frequency content
+
+    '''
+    assert len(ref_freqs) == len(dis_freqs), "Mismatched number of trials"
+
+    trial_pairs = []
+    for i in range(len(ref_freqs)):
+        curr_r = ref_freqs[i]
+        curr_d = dis_freqs[i]
+        # Find next trial with complementary frequency content
+        for j in range(i+1,len(ref_freqs)):
+            next_r = ref_freqs[j]
+            next_d = dis_freqs[j]
+            if (next_r != curr_r).all() and (next_d != curr_d).all() and j-i <= max_trial_distance:
+                trial_pairs.append([i,j])
+                break
```

Comment:
why do you need the break here?

Reply:
Good catch! I had more code underneath this in the past and forgot to remove the `break`

Reply:
okay!

Reply:
Actually, I reviewed this function and wrote some tests and there was a reason I added the break here. It's related to how many pairs you're okay with a single trial contributing to (in experiments where it's not set up for the ref/dis freqs to alternate on every trial), and it's a bit of a subjective decision. Maybe we can talk about it in person?